### DISCUSSION TITLE
Can you combine a controller and a repeater/gateway in the same board.

### DISCUSSION THREAD

**Discussion by PerRieland** [timmbogner/Farm-Data-Relay-System#170] [Q&A]
Is it possible to have a controller that can function as a controller and a repeater/gateway ?

I have had _no_ success in the following setup. 

<=WiFi=> **MQTT** <=LoRa=> **Controller1** <=ESPNOW=> **Controller2**

Controller1 must act as an irrigation controller _and_ repeater/gateway.
Controller2 must act as an irrigation controller.

I have a long distance between the MQTT Gateway and Controller1 and here is LoRa perfect. This setup works perfectly.
Now I want to extend the setup with one more irrigation controller.
Controller1 and Controller2 have a short distance. Here ESPNOW will be a good choice for wireless communication between Controller1 and Controller2. 

I’m using [WIFI LoRa 32 (V3) – Heltec Automation](https://heltec.org/project/wifi-lora-32-v3/) board for MQTT and Controller1, which have both LoRa and WiFi. Controller2 is a Wemos mini D1 8266.

Nothing shows up in serial monitor on Controller1.

Here is the config files
**MQTT**: 
//Addresses
#define UNIT_MAC           0x00  // The address of this gateway

#define ESPNOW_NEIGHBOR_1  0x00  // Address of ESP-NOW neighbor 1
#define ESPNOW_NEIGHBOR_2  0x00  // Address of ESP-NOW neighbor 2
#define LORA_NEIGHBOR_1    0x00  // Address of LoRa neighbor 1
#define LORA_NEIGHBOR_2    0x00  // Address of LoRa neighbor 2

// Interfaces
#define USE_LORA
#define USE_WIFI  

// Routing
// Options: sendESPNowNbr(1 or 2); sendESPNowPeers(); sendLoRaNbr(1 or 2); broadcastLoRa(); sendSerial(); sendMQTT();
#define ESPNOWG_ACT    
#define LORAG_ACT      sendMQTT();
#define SERIAL_ACT    
#define MQTT_ACT       broadcastLoRa(); sendSerial();
#define INTERNAL_ACT  
#define ESPNOW1_ACT  
#define ESPNOW2_ACT                    
#define LORA1_ACT      
#define LORA2_ACT

**Controller1**:
//Addresses
#define UNIT_MAC           0x03  // The address of this gateway

#define ESPNOW_NEIGHBOR_1  0x00  // Address of ESP-NOW neighbor 1
#define ESPNOW_NEIGHBOR_2  0x07  // Address of ESP-NOW neighbor 2
#define LORA_NEIGHBOR_1    0x00  // Address of LoRa neighbor 1
#define LORA_NEIGHBOR_2    0x99  // Address of LoRa neighbor 2

// Interfaces
#define USE_ESPNOW  
#define USE_LORA

// Routing
// Options: sendESPNowNbr(1 or 2); sendESPNowPeers(); sendLoRaNbr(1 or 2); broadcastLoRa(); sendSerial(); sendMQTT();
#define ESPNOWG_ACT    sendSerial(); sendESPNowNbr(1);
#define LORAG_ACT      sendSerial(); sendLoRaNbr(1); sendESPNowNbr(2);
#define SERIAL_ACT    
#define MQTT_ACT          
#define INTERNAL_ACT   sendSerial(); sendLoRaNbr(1);
#define ESPNOW1_ACT    
#define ESPNOW2_ACT    sendSerial(); sendLoRaNbr(1);                
#define LORA1_ACT      sendSerial(); sendESPNowNbr(2); broadcastLoRa();
#define LORA2_ACT      

**Controller2**:
#define UNIT_MAC      0x07  // The address of this gateway
#define READING_ID    1007   //Unique ID for this sensor
#define GTWY_MAC      0x03 //Using a repeater with address 0x03.  

#define USE_ESPNOW
#define FDRS_DEBUG

**Comment by timmbogner** [comment]
Hey Per,
The intended configuration for your scenario is pictured below. Gateway 0x03 needs a strong solar panel to remain running and routing packets, allowing the controllers to run more independently. Controllers can spend most of their time asleep, then wake up and ask the front end for a command. This sleeping controller setup allows you to use a much smaller panel, smaller battery, and inexpensive equipment for the controller. I don't have a good example built for it yet, but it involves a flow for Node-RED that listens for "I'm awake!" messages and sends any commands that it may have missed.
![IMG_20230508_095329411](https://user-images.githubusercontent.com/64260873/236865857-644ef179-bd61-4f55-af96-c36371f12498.jpg)

The concept of gateways as controllers isn't my preferred route for the project. The intent was to make modular access points so that devices could worry less about power and networking. FDRS puts all the burden of staying alive and stable on the gateway, so that the sensor/controller node can be as simple or complicated as needed without jeopardizing network stability. I realize this and other aspects of FDRS sometimes mean you need an extra ESP or two, but I think the extra stability is worth it!

There is an ongoing project called Rezodo on Hackaday by a friend who modified FDRS to what  you describe, and I encourage you to check that out for a LOT of reasons (cool hardware!).

**Reply by timmbogner** [reply]
Okay, I got some coffee and decided it's just easier to post the two gateway configs:
UART Gateway (0x01):
``` cpp
//  FARM DATA RELAY SYSTEM
//
//  GATEWAY CONFIGURATION

//Addresses
#define UNIT_MAC           0x01  // The address of this gateway

#define ESPNOW_NEIGHBOR_1  0x00  // Address of ESP-NOW neighbor #1
#define ESPNOW_NEIGHBOR_2  0x02  // Address of ESP-NOW neighbor #2
#define LORA_NEIGHBOR_1    0x00  // Address of LoRa neighbor #1
#define LORA_NEIGHBOR_2    0x03  // Address of LoRa neighbor #2

// Interfaces
//#define USE_ESPNOW  
#define USE_LORA
#define USE_WIFI  // Will cause errors if used with ESP-NOW. Use a serial link instead!
//#define USE_ETHERNET

// Routing
// Options: sendESPNowNbr(1 or 2); sendESPNowPeers(); sendLoRaNbr(1 or 2); broadcastLoRa(); sendSerial(); sendMQTT();
#define ESPNOWG_ACT    sendSerial(); sendMQTT();
#define LORAG_ACT      sendSerial(); sendMQTT();
#define SERIAL_ACT     sendESPNowNbr(2); sendESPNowPeers(); sendLoRaNbr(2); broadcastLoRa(); 
#define MQTT_ACT       sendESPNowNbr(2); sendESPNowPeers(); sendLoRaNbr(2); broadcastLoRa();   
#define INTERNAL_ACT   sendSerial(); sendMQTT();
#define ESPNOW1_ACT    
#define ESPNOW2_ACT    sendSerial(); sendMQTT();               
#define LORA1_ACT      
#define LORA2_ACT      sendSerial(); sendMQTT();

#define RADIOLIB_MODULE SX1262  // ESP32 SX1262 (HELTEC)
#define LORA_SS 8
#define LORA_RST 12
#define LORA_DIO 14
#define LORA_BUSY 13
#define USE_SX126X

#define LORA_TXPWR 17   // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))

//#define CUSTOM_SPI
#define LORA_SPI_SCK  5
#define LORA_SPI_MISO 19
#define LORA_SPI_MOSI 27

#define FDRS_DEBUG     // Enable USB-Serial debugging

// OLED -- Displays console debugging messages on an SSD1306 I²C OLED
///#define USE_OLED    
#define OLED_HEADER "FDRS"
#define OLED_PAGE_SECS 30
#define OLED_SDA 4
#define OLED_SCL 15
#define OLED_RST 16

// UART data interface pins (if available)
#define RXD2 14
#define TXD2 15

//#define USE_LR  // Use ESP-NOW LR mode (ESP32 only)

// WiFi and MQTT Credentials  -- These will override the global settings
//#define WIFI_SSID   "Your SSID"  
//#define WIFI_PASS   "Your Password"

//#define MQTT_ADDR   "192.168.0.8"
//#define MQTT_PORT   1883 // Default MQTT port is 1883
//#define MQTT_AUTH   //Enable MQTT authentication 
//#define MQTT_USER   "Your MQTT Username"
//#define MQTT_PASS   "Your MQTT Password"
```

Repeater (0x03):
``` cpp
//  FARM DATA RELAY SYSTEM
//
//  GATEWAY CONFIGURATION

//Addresses
#define UNIT_MAC           0x03  // The address of this gateway

#define ESPNOW_NEIGHBOR_1  0x00  // Address of ESP-NOW neighbor #1
#define ESPNOW_NEIGHBOR_2  0x00  // Address of ESP-NOW neighbor #2
#define LORA_NEIGHBOR_1    0x01  // Address of LoRa neighbor #1
#define LORA_NEIGHBOR_2    0x05  // Address of LoRa neighbor #2

// Interfaces
#define USE_ESPNOW  
#define USE_LORA
//#define USE_WIFI  // Will cause errors if used with ESP-NOW. Use a serial link instead!
//#define USE_ETHERNET

// Routing
// Options: sendESPNowNbr(1 or 2); sendESPNowPeers(); sendLoRaNbr(1 or 2); broadcastLoRa(); sendSerial(); sendMQTT();
#define ESPNOWG_ACT    sendLoRaNbr(1);
#define LORAG_ACT      sendLoRaNbr(1);
#define SERIAL_ACT     sendESPNowPeers();
#define MQTT_ACT          
#define INTERNAL_ACT   sendLoRaNbr(1);
#define ESPNOW1_ACT    
#define ESPNOW2_ACT                    
#define LORA1_ACT      sendLoRaNbr(2); broadcastLoRa(); sendESPNowPeers();
#define LORA2_ACT      sendLoRaNbr(1);

#define RADIOLIB_MODULE SX1262  // ESP32 SX1262 (HELTEC)
#define LORA_SS 8
#define LORA_RST 12
#define LORA_DIO 14
#define LORA_BUSY 13
#define USE_SX126X

#define LORA_TXPWR 17   // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))

//#define CUSTOM_SPI
#define LORA_SPI_SCK  5
#define LORA_SPI_MISO 19
#define LORA_SPI_MOSI 27

#define FDRS_DEBUG     // Enable USB-Serial debugging

// OLED -- Displays console debugging messages on an SSD1306 I²C OLED
///#define USE_OLED    
#define OLED_HEADER "FDRS"
#define OLED_PAGE_SECS 30
#define OLED_SDA 4
#define OLED_SCL 15
#define OLED_RST 16

// UART data interface pins (if available)
#define RXD2 14
#define TXD2 15

//#define USE_LR  // Use ESP-NOW LR mode (ESP32 only)

// WiFi and MQTT Credentials  -- These will override the global settings
//#define WIFI_SSID   "Your SSID"  
//#define WIFI_PASS   "Your Password"

//#define MQTT_ADDR   "192.168.0.8"
//#define MQTT_PORT   1883 // Default MQTT port is 1883
//#define MQTT_AUTH   //Enable MQTT authentication 
//#define MQTT_USER   "Your MQTT Username"
//#define MQTT_PASS   "Your MQTT Password"
```

Controllers:
``` cpp
//  FARM DATA RELAY SYSTEM
//
//  Sensor Configuration

#include <fdrs_globals.h>

#define READING_ID    56   //Unique ID for this sensor
#define GTWY_MAC      0x03 //Address of the nearest gateway

#define USE_ESPNOW
//#define USE_LORA
//#define DEEP_SLEEP
//#define POWER_CTRL    14
#define FDRS_DEBUG

// LoRa Configuration
#define RADIOLIB_MODULE SX1276
#define LORA_SS    18
#define LORA_RST   14
#define LORA_DIO   26
#define LORA_BUSY  33
//#define USE_SX126X

#define LORA_TXPWR 17   // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))
#define LORA_ACK        // Request LoRa acknowledgment.
```

**Reply by PerRieland** [reply]
Thank you. I did some of the things in the configuration files but had no success. Maybe I missed something.
I will definitely go through your configuration files and try again.
